### PR TITLE
Idris: Wrap with IDRIS_CC

### LINKED ISSUE #42785: idris's wrapper should set IDRIS_CC to reduce the risk of accidentally using the wrong CC version
Idris requires a C compiler to function. To find a C compiler, it first checks IDRIS_CC, as implemented here: https://github.com/idris-lang/Idris-dev/blob/master/src/IRTS/System.hs#L63 and then searches PATH.

The current wrapper for the "idris" compiler binary appends /nix/store/...-gcc-wrapper/bin to the PATH. This unfortunately makes idris quite fragile: When my PATH has some bad program named "cc" on it, such as some ancient compiler that I've installed for whatever reason, idris will break. It's also a fairly unnecessary impurity.

We should set IDRIS_CC in the wrapper instead, when it is not already set. This allows the user to still override the C compiler that is used if they want to, while significantly lowering the chance that things will be unexpectedly broken.

### PR DESCRIPTION
**Pull request by infinisil**
###### Motivation for this change
Previously idris used the C compiler from PATH for the C backend, which
means that the results and whether it even succeeds can vary between
systems (e.g. if a Nix-built Idris was used on a super old Linux system,
the cc installed there might not even work for Idris' C).

To make this more predictable, this commit sets the IDRIS_CC env var,
which Idris will prefer over searching in PATH, to a Nix-provided gcc
executable, given that it is not already set, so it's still possible to
override.

This closes #42785

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by Infinisil** [comment]
Ping @brainrape

File: pkgs/development/idris-modules/idris-wrapper.nix [hunk 1]
```diff
@@ -1,14 +1,12 @@
...
+  paths = [ idris-no-deps ];
   buildInputs = [ makeWrapper ];
-  meta.platforms = idris.meta.platforms;
   postBuild = ''
     wrapProgram $out/bin/idris \
-      --suffix PATH : ${ stdenv.lib.makeBinPath path } \
-      --suffix LIBRARY_PATH : ${stdenv.lib.makeLibraryPath lib}
-      '';
-  }
+      --run 'export IDRIS_CC=''${IDRIS_CC:-${lib.getBin gcc}/bin/gcc}' \
```
Comment (brainrape) [inline pkgs/development/idris-modules/idris-wrapper.nix]: Can we use `makeWrapper`'s `--set` parameter here? See:
https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/setup-hooks/make-wrapper.sh#L16

**Comment by brainrape** [comment]
Thanks @Infinisil this looks good and useful, except the above.

File: pkgs/development/idris-modules/idris-wrapper.nix (hunk 1 above)
Comment (Infinisil) [reply pkgs/development/idris-modules/idris-wrapper.nix]: Unfortunately not, `--set` really sets it equal to exactly the string you give it, can't expand any variables (it uses bash's `${parameter@operator}` "Parameter transformation" with operator=Q, see `man bash`). `--run` is the next best way.

Comment (brainrape) [reply pkgs/development/idris-modules/idris-wrapper.nix]: I see. All good then.

**Comment by brainrape** [comment]
Good to merge.

**Pull request merged**